// @dotenv-run/webpack/dist/esm/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { env } from "@dotenv-run/core";
import { DefinePlugin } from "webpack";
var DotenvRunPlugin = /** @class */ (function () {
    function DotenvRunPlugin(options, ssr) {
        if (ssr === void 0) { ssr = false; }
        this.ssr = ssr;
        this.raw = {};
        this.full = {};
        this.stringified = {};
        var _a = env(options), full = _a.full, stringified = _a.stringified, raw = _a.raw;
        this.raw = raw;
        this.full = full;
        this.stringified = stringified;
    }
    DotenvRunPlugin.prototype.apply = function (compiler) {
        var definePlugin = new DefinePlugin(this.ssr
            ? __assign({}, this.full) : {
            "process.env": this.stringified,
            "import.meta.env": this.stringified,
            "import.meta.env.NODE_ENV": JSON.stringify(process.env.NODE_ENV),
        });
        definePlugin.apply(compiler);
    };
    return DotenvRunPlugin;
}());
export { DotenvRunPlugin };
export default DotenvRunPlugin;
