// @dotenv-run/webpack/dist/cjs/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DotenvRunPlugin = void 0;
var core_1 = require("@dotenv-run/core");
var webpack_1 = require("webpack");
var DotenvRunPlugin = /** @class */ (function () {
    function DotenvRunPlugin(options, ssr) {
        if (ssr === void 0) { ssr = false; }
        this.ssr = ssr;
        this.raw = {};
        this.full = {};
        this.stringified = {};
        var _a = (0, core_1.env)(options), full = _a.full, stringified = _a.stringified, raw = _a.raw;
        this.raw = raw;
        this.full = full;
        this.stringified = stringified;
    }
    DotenvRunPlugin.prototype.apply = function (compiler) {
        var definePlugin = new webpack_1.DefinePlugin(this.ssr
            ? __assign({}, this.full) : {
            "process.env": this.stringified,
            "import.meta.env": this.stringified,
            "import.meta.env.NODE_ENV": JSON.stringify(process.env.NODE_ENV),
        });
        definePlugin.apply(compiler);
    };
    return DotenvRunPlugin;
}());
exports.DotenvRunPlugin = DotenvRunPlugin;
exports.default = DotenvRunPlugin;
