// @ngx-env/builder/dist/builders/unit-test/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeWithEnv = void 0;
const architect_1 = require("@angular-devkit/architect");
const build_1 = require("@angular/build");
const core_1 = require("@dotenv-run/core");
const rxjs_1 = require("rxjs");
const get_environment_1 = require("../utils/get-environment");
const validate_prepare_context_1 = require("../utils/validate-prepare-context");
const project_1 = require("../utils/project");
const executeWithEnv = (options, context) => {
    let _options = {
        ...options,
        buildTarget: options.buildTarget ?? "::development", // as in normalizeOptions
        // if browsers is set to an empty array, it breaks the unit test builder
        // if browsers is an empty array, set it to undefined, otherwise set it to the browsers array
        browsers: options.browsers?.length > 0 ? options.browsers : undefined,
    };
    return (0, rxjs_1.combineLatest)([
        (0, validate_prepare_context_1.validateAndPrepareBuildContext)(_options.buildTarget, context),
        (0, project_1.getProjectCwd)(context),
    ]).pipe((0, rxjs_1.switchMap)(([buildTarget, cwd]) => {
        const { full } = (0, core_1.env)({
            ...buildTarget.ngxEnv,
            runtime: false, // Runtime is not supported in unit tests
            cwd,
            environment: (0, get_environment_1.getEnvironment)(buildTarget.configuration),
        });
        context.getTargetOptions = async () => ({
            ...buildTarget.options,
            define: full,
        });
        if (buildTarget.ngxEnv?.runtime) {
            console.warn("NOTE: @ngx-env/builder runtime option is ignored in unit tests.");
        }
        console.warn("WARNING: @ngx-env/builder vitest support is experimental.");
        return (0, build_1.executeUnitTestBuilder)(_options, context);
    }), (0, rxjs_1.catchError)((e) => {
        console.error(e);
        return (0, build_1.executeUnitTestBuilder)(_options, context);
    }));
};
exports.executeWithEnv = executeWithEnv;
exports.default = (0, architect_1.createBuilder)(exports.executeWithEnv);
